tracer/buffer.go

```diff
@@ -1,60 +1,115 @@
 package tracer
 
 import (
-	"math/rand"
+	"fmt"
 	"sync"
 )
 
 const (
-	spanBufferDefaultMaxSize = 10000
+	// traceBufferDefaultMaxSize is the maximum number of spans we keep in memory.
+	// This is to avoid memory leaks, if above that value, spans are randomly
+	// dropped and ignore, resulting in corrupted tracing data, but ensuring
+	// original program continues to work as expected.
+	traceBufferDefaultMaxSize = 10000
 )
 
-// spansBuffer is a threadsafe buffer for spans.
-type spansBuffer struct {
-	lock    sync.Mutex
+type traceBuffer struct {
+	// spans is a traceBuffer containing all the spans for this trace.
+	// The reason we don't use a channel here, is we regularly need
+	// to walk the array to find out if it's done or not.
 	spans   []*Span
 	maxSize int
-}
 
-func newSpansBuffer(maxSize int) *spansBuffer {
+	traceChan chan<- []*Span
+	errChan   chan<- error
+
+	sync.RWMutex
+}
 
-	// small sanity check on the max size.
+func newTraceBuffer(traceChan chan<- []*Span, errChan chan<- error, maxSize int) *traceBuffer {
 	if maxSize <= 0 {
-		maxSize = spanBufferDefaultMaxSize
+		maxSize = traceBufferDefaultMaxSize
+	}
+	return &traceBuffer{
+		traceChan: traceChan,
+		errChan:   errChan,
+		maxSize:   maxSize,
+	}
+}
+
+func (tb *traceBuffer) doPush(span *Span) {
+	tb.Lock()
+	defer tb.Unlock()
+
+	// if traceBuffer is full, forget span
+	if len(tb.spans) >= tb.maxSize {
+		select {
+		case tb.errChan <- fmt.Errorf("[TODO:christian] exceed traceBuffer size"):
+		default: // if channel is full, drop & ignore error, better do this than stall program
+		}
+		return
+	}
+	// if there's a trace ID mismatch, ignore span
+	if len(tb.spans) > 0 && tb.spans[0].TraceID != span.TraceID {
+		select {
+		case tb.errChan <- fmt.Errorf("[TODO:christian] trace ID mismatch"):
+		default: // if channel is full, drop & ignore error, better do this than stall program
+		}
+		return
 	}
 
-	return &spansBuffer{maxSize: maxSize}
+	tb.spans = append(tb.spans, span)
 }
 
-func (sb *spansBuffer) Push(span *Span) {
-	sb.lock.Lock()
-	if len(sb.spans) < sb.maxSize {
-		sb.spans = append(sb.spans, span)
-	} else {
-		idx := rand.Intn(sb.maxSize)
-		sb.spans[idx] = span
+func (tb *traceBuffer) Push(span *Span) {
+	if tb == nil {
+		return
 	}
-	sb.lock.Unlock()
+	tb.doPush(span)
 }
 
-func (sb *spansBuffer) Pop() []*Span {
-	sb.lock.Lock()
-	defer sb.lock.Unlock()
+func (tb *traceBuffer) flushable() bool {
+	tb.RLock()
+	defer tb.RUnlock()
 
-	if len(sb.spans) == 0 {
-		return nil
+	if len(tb.spans) == 0 {
+		return false
 	}
 
-	// FIXME[matt] on rotation, we could re-use the slices and spans here
-	// and avoid re-allocing.
-	spans := sb.spans
-	sb.spans = nil
+	for _, span := range tb.spans {
```

Comment:
instead of looping each time you try to do a `doFlush`, can't you store an `int` that counts the number of finished spans? even if looping here is not so expensive in the average case, we still have a loop and a `RLock()`. So to check if a buffer is flushable, you may:
* when you add a `Span` in `doPush()`, increase the `traceBuffer.finishedSpan` counter (when a `Span` is finished, its state is immutable so it's safe to always increase that number)
* when you want to check if a buffer is finished, you check `if traceBuffer.finishedSpan == len(traceBuffer.spans)`

makes sense?